### PR TITLE
Add an editor-finder plugin

### PR DESCRIPTION
**Pull request by KrzysztofSikoraCodete**
Refactor tracker.ts

Separated method find from tracker handler.
Created stand alone editor finder.

The plugin must be separated, instead of create the interface and token into tracker.ts, because if we don't do that will not be available in whole scope and will be a conflict with dependency cycle.

### CODE REVIEW

File: src/service.ts [hunk 1]
```diff
@@ -234,7 +234,25 @@ export class DebuggerService implements IDebugger, IDisposable {
         );
       });
     }
-
+    const unassociatedBreakpoints = (
```
Comment (jtpio) [inline src/service.ts:237]: Don't we want to keep the change related to the breakpoints in a separate PR, and keep this PR focused on the refactor of the tracker? (extracting the "find in" logic into a separate plugin)

What do you think?

Comment (KrzysztofSikoraCodete) [reply src/service.ts:237]: It makes sense. I will remove this method from this PR.

Comment (KrzysztofSikoraCodete) [reply src/service.ts]: I've just removed.

File: src/handlers/tracker.ts [hunk 2]
```diff
@@ -360,21 +204,10 @@ export namespace TrackerHandler {
...
-     * An optional tracker for consoles.
-     */
-    consoleTracker?: IConsoleTracker;
-
     /**
-     * An optional tracker for file editors.
+     * An optional editor finder
      */
-    editorTracker?: IEditorTracker;
+    editorFinder?: IDebuggerEditorFinder;
```
Comment (afshin) [inline src/handlers/tracker.ts:210]: This should not be optional I think.

File: src/index.ts [hunk 3]
```diff
@@ -240,30 +242,51 @@ const notebooks: JupyterFrontEndPlugin<void> = {
...
+      editorFinder
+    });
+  }
+};
+
+/**
+ * A plugin that tracks editors, console and file editors used for debugging.
+ */
+const finder: JupyterFrontEndPlugin<EditorFinder> = {
+  id: '@jupyterlab/debugger:tracker',
```
Comment (jtpio) [inline src/index.ts:270]: Should this be renamed to?

```suggestion
  id: '@jupyterlab/debugger:finder',
```

File: src/editor-finder.ts [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+import { IEditorTracker } from '@jupyterlab/fileeditor';
+import { INotebookTracker } from '@jupyterlab/notebook';
+import { chain, each } from '@lumino/algorithm';
+import { IIterator } from '@lumino/algorithm/lib/iter';
+import { Token } from '@lumino/coreutils';
+import { IDisposable } from '@lumino/disposable';
+import { Signal } from '@lumino/signaling';
+import { IDebugger } from './tokens';
+
+export class EditorFinder implements IDisposable, IDebuggerEditorFinder {
```
Comment (jtpio) [inline src/editor-finder.ts:21]: How about adding docstrings for the class and the constructor?

File: src/editor-finder.ts [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+import { IEditorTracker } from '@jupyterlab/fileeditor';
+import { INotebookTracker } from '@jupyterlab/notebook';
+import { chain, each, IIterator } from '@lumino/algorithm';
+import { Token } from '@lumino/coreutils';
+import { IDisposable } from '@lumino/disposable';
+import { Signal } from '@lumino/signaling';
+import { IDebugger } from './tokens';
+
+export class EditorFinder implements IDisposable, IDebuggerEditorFinder {
+  constructor(options: EditorFinder.IOptions) {
```
Comment (jtpio) [inline src/editor-finder.ts:21]:
```suggestion
/**
  * A class to find instances of code editors across notebook, console and files widgets
  */
 export class EditorFinder implements IDisposable, IDebuggerEditorFinder {
  /**
   * Instantiate a new EditorFinder.
   * @param options The instantiation options for a EditorFinder.
   */
  constructor(options: EditorFinder.IOptions) {
```

Comment (jtpio) [reply src/editor-finder.ts:21]: Trying to add them with an inline suggestion :crossed_fingers:

**Comment by jtpio** [comment]
Test failure might be relevant because of the way the model is now instantiated.

**Comment by jtpio** [comment]
Tests are passing locally, but they are timing out on CI.

**Comment by jtpio** [comment]
Correction: the tests also fail locally now after upgrading the dependencies.

Among the updates, two seem worth having a look:

```
  jupyterlab_server                              1.1.3-py_0 --> 1.1.5-py_0
  xeus                                   0.23.14-h4d8c418_0 --> 0.23.15-h4d8c418_0
```

**Comment by jtpio** [comment]
Debugging this locally to find which version to pin to for now on CI and in the dev setup.

**Comment by jtpio** [comment]
Looks like it's related to `xeus=0.23.15`. Downgrading to `xeus=0.23.14` fixes the tests locally and on CI.

cc @JohanMabille do you know what changed in `xeus=0.23.15`? It looks like there are some issues starting the kernels:

```
 RUNS  src/session.spec.ts
[I 11:26:17.752 JestApp] KernelRestarter: restarting kernel (1/5), new random ports
 RUNS  src/session.spec.ts
ERROR: received bad message: No such file or directory
 RUNS  src/session.spec.ts
[I 11:26:54.732 JestApp] Kernel started: 668d6dc5-ee1e-4e82-8f66-2402a9215fda
[I 11:26:54.778 JestApp] Kernel started: f32df121-271e-4f99-8a20-5e94f4d8d6e1
 RUNS  src/session.spec.ts
[I 11:26:57.777 JestApp] KernelRestarter: restarting kernel (1/5), new random ports
 RUNS  src/session.spec.ts
[W 11:27:10.985 JestApp] Timeout waiting for kernel_info reply from 6358d30a-9ae4-4ae5-90db-8b32298474da
```  

- Failed run: https://github.com/jupyterlab/debugger/pull/441/checks?check_run_id=712434296
- Succeeded run (after pinning to `xeus=0.23.14`): https://github.com/jupyterlab/debugger/pull/441/checks?check_run_id=712506816

**Comment by jtpio** [comment]
We can keep the pin until we update the tests to use the `JupyterServer` from `@jupyterlab/testutils` (https://github.com/jupyterlab/debugger/issues/445).

**Comment by jtpio** [comment]
For the record, this is reproducible locally with:

```bash
conda install -c conda-forge xeus=0.23.15
XEUS_LOG=1 jlpm run build && jlpm run test
```

**Comment by jtpio** [comment]
Let's merge and track the version pin separately.

**Comment by afshin** [comment]
Good catch, Jeremy! Cheer!

**Pull request merged**